client/app/pages/alert/Alert.jsx

```diff
@@ -383,14 +194,20 @@ export default function init(ngModule) {
   ngModule.component('alertPage', react2angular(AlertPage));
 
   return routesToAngularRoutes([
+    {
+      path: '/alerts/new',
+      title: 'New Alert',
+      mode: MODES.NEW,
```

Comment:
Why not have each path render its own component? What's the purpose of having a single top level page component? To share the callback functions?

Reply:
Yes. Less code duplication
Also allows `alert`, `query` and `queryResult` to be props instead of state, simplifying the page components.

Reply:
Also allows this, which imo is more elegant than redirection to view page:
https://github.com/getredash/redash/blob/9aeccb4d8fc251bc6110441539c7a7415ab038ae/client/app/pages/alert/Alert.jsx#L68-L71

Reply:
You could use composition instead:

```
<AlertPage ...>
    <EditAlertPage ...>
</AlertPage>
```

*But* the ability to change states without making API request is a win. In the long run, we should find a way to do this without as this pattern isn't always applicable. But here it's fine.

Reply:
> _But_ the ability to change states without making API request is a win.

Are you sure that's happening here? Changing Alert states mandates url changes which causes page to reload (`navigateTo`).

Reply:
@arikfr it was an easy modification https://github.com/getredash/redash/pull/4183/commits/e8654600daff0c4f6e9a08af017315a482130bb4